aten/src/ATen/native/sparse/SparseFactories.cpp

```diff
@@ -0,0 +1,171 @@
+            scalar_t* diag_data = reinterpret_cast<scalar_t*>(diagonals_bytes);
+            int64_t offset = *reinterpret_cast<int64_t*>(offsets_bytes);
+            int64_t nnz_len = *reinterpret_cast<int64_t*>(nnz_per_diag_bytes);
+            int64_t nnz_prefix = *reinterpret_cast<int64_t*>(nnz_prefix_bytes);
+
+            int64_t col_offset = std::max<int64_t>(offset, 0);
+            int64_t row_offset = col_offset - offset;
+
+            for (int64_t i = 0; i < nnz_len; ++i) {
+              values_data[nnz_prefix + i] =
+                  diag_data[(col_offset + i) * diag_stride];
+            }
+
+            std::copy(
+                index_buffer_data + row_offset,
+                index_buffer_data + row_offset + nnz_len,
+                row_indices_data + nnz_prefix);
+            std::copy(
+                index_buffer_data + col_offset,
+                index_buffer_data + col_offset + nnz_len,
+                col_indices_data + nnz_prefix);
+
+            diagonals_bytes += diagonals_stride;
+            offsets_bytes += offsets_stride;
+            nnz_per_diag_bytes += nnz_per_diag_stride;
+            nnz_prefix_bytes += nnz_prefix_stride;
+          }
+        };
+        iter.for_each(loop);
+      });
+}
+
+// Check offsets for duplicates, and out-of-bounds diagonals
+// While checking offsets, compute nnz per diagonal array, prefix sum array, and
+// nnz total for later use
+void _spdiags_setup_cpu(
+    TensorIterator& iter,
+    int64_t n_row_out,
+    int64_t n_col_in,
+    int64_t n_col_out,
+    int64_t& nnz_out) {
+  const int64_t min_col_in_out = std::min(n_col_in, n_col_out);
+  auto loop = [&](char** data, const int64_t* strides, int64_t n) {
+    auto* nnz_per_diag_bytes = data[0];
+    auto* offsets_bytes = data[1];
+    std::set<int64_t> seen_offsets;
+    for (int64_t d = 0; d < n; ++d) {
+      auto* nnz_per_diag_data = reinterpret_cast<int64_t*>(nnz_per_diag_bytes);
+      auto offset = *reinterpret_cast<int64_t*>(offsets_bytes);
+      TORCH_CHECK(
+          seen_offsets.insert(offset).second,
+          "spidags(): Offset array contains duplicate values");
+      TORCH_CHECK(
+          ((-1 * n_row_out) < offset) && (offset < n_col_out),
+          "spdiags(): Diagonal ",
+          offset,
+          " does not exist in output shape (",
+          n_row_out,
+          ",",
+          n_col_out,
+          "). Valid offsets for this shape: [",
+          (-n_row_out) + 1,
+          ",",
+          n_col_out - 1,
+          "]");
+      if (offset >= 0) {
+        *nnz_per_diag_data =
+            std::max<int64_t>(std::min(min_col_in_out - offset, n_row_out), 0);
+      } else {
+        *nnz_per_diag_data =
+            std::max<int64_t>(std::min(min_col_in_out, n_row_out + offset), 0);
+      }
+      nnz_out += *nnz_per_diag_data;
+      nnz_per_diag_bytes += strides[0];
+      offsets_bytes += strides[1];
+    }
+  };
+  iter.for_each(loop);
```

Comment:
race condition: `nnz_out` is potentially shared among different threads that try to modify it without any synchronization.